src/mlpack/methods/ann/rbm/spike_slab_rbm_policy.hpp

```diff
@@ -0,0 +1,224 @@
+ */
+#ifndef MLPACK_METHODS_ANN_RBM_SPIKE_SLAB_RBM_POLICY_HPP
+#define MLPACK_METHODS_ANN_RBM_SPIKE_SLAB_RBM_POLICY_HPP
+
+#include <mlpack/core.hpp>
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/ann/activation_functions/softplus_function.hpp>
+
+namespace mlpack {
+namespace ann {
+
+class SpikeSlabRBMPolicy
+{
+ public:
+  // Intialise the visible and hiddenl layer of the network
+  SpikeSlabRBMPolicy(const size_t visibleSize,
+      const size_t hiddenSize,
+      const size_t poolSize,
+      const double slabPenalty, 
+      double radius);
+
+  // Reset function
+  void Reset();
+
+  /**
+   * Free energy of the spike and slab variable
+   * the free energy of the ssRBM is given my
+   * $v^t$$\Delta$v - $\sum_{i=1}^N$ 
+   * $\log{ \sqrt{\frac{(-2\pi)^K}{\prod_{m=1}^{K}(\alpha_i)_m}}}$ -
+   * $\sum_{i=1}^N \log(1+\exp( b_i +
+   * \sum_{m=1}^k \frac{(v(w_i)_m^t)^2}{2(\alpha_i)_m})$
+   *
+   * @param input the visible layer
+   */ 
+  double FreeEnergy(arma::mat&& input);
+
+  /**
+   * Evaluate function is used by the Optimizer to 
+   * find the perfomance of the network on the currentInput
+   */
+  double Evaluate(arma::mat& /*predictors*/, size_t /*i*/);
+
+  /**
+   * Positive Phase calculates the gradient on the
+   * given data point. 
+   *
+   *
+   * @param input the visible input
+   * @param output the computed gradient
+   */
+  void PositivePhase(arma::mat&& input, arma::mat&& gradient);
+
+  /**
+   * Negative Phase calculates the gradient on the
+   * negative sample obtained for the given data point
+   * using gibbs sampling. 
+   *
+   * @param input the visible input
+   * @param output the computed gradient
+   */
+  void NegativePhase(arma::mat&& negativeSamples, arma::mat&& gradient);
+
+  /**
+   * Visible Mean function calculates the mean of the
+   * normal distribution of P(v| s,h).
+   * Where the mean is given by \Lambda^{-1} \sum_{i=1}^N W_i * s_i * h_i 
+   * 
+   * @param input consists of spike and slab variables
+   * @param output the mean of the of the Normal distribution
+   */
+  void VisibleMean(arma::mat&& input, arma::mat&& output);
+
+  /**
+   * Hidden Mean function calculates the
+   * normal distribution of P(s|v,h).
+   * Where the mean is given by h_i*\alpha^{-1}*W_i^T*v 
+   * variance is givenby \alpha^{-1}
+   * 
+   * @param input consists of visible input.
+   * @param output consits of the spike samples and slab samples
+   */
+  void HiddenMean(arma::mat&& input, arma::mat&& output);
+  
+  /**
+   * Sample Visible function sample 
+   * the visible outputs from the normal distribution with
+   * mean \Lambda^{-1} \sum_{i=1}^N W_i * s_i * h_i and 
+   * variance \Lambda^{-1}
+   * 
+   * @param input consists of spike and slab variables
+   * @param output consits of visible layer.
+   * @param norm norm used for rejection sampling
+   */
+  void SampleVisible(arma::mat&& input, arma::mat&& output, size_t norm = 2);
```

Comment:
First, armadillo supports `"inf"` and `"-inf"` norms, you can't use them since the datatype is `size_t`. Second, it is better to use the same API for both `BinaryRBMPolicy` and `SpikeSlabRBMPolicy`. Third, did you try other norms? I suggested that but later I realized that some norms cause failures in `SampleVisible()` and the model may fall into instability. So, right now I am not sure that this is really needed. Sorry, that I didn't say that earlier.